### PR TITLE
Issue #4719 Keep previous charencoding if not set in content type

### LINKED ISSUE #4719: ContentType with no char encoding should use previous char encoding
**jetty-10.0.x**

Servlet tck test com/sun/ts/tests/servlet/api/javax_servlet/servletresponse/URLClient.java.getContentTypeNull2Test.

If Response.setContentType() is called with a string that has no char encoding, any previously set char encoding should be used rather than the default char set for the new content type.
Eg 
response.setContentType("foo/bar; charset=some-charset")
the charset encoding should now be some-charset
response.setContentType("text/xml")
the charset encoding should still be some-charset

### PR DESCRIPTION
**Pull request by janbartel**
Closes #4719

### CODE REVIEW

**Review by gregw — APPROVED** [review]

**Review by joakime — COMMENTED** [review]
Questions about this one.

Does this preservation persist through dispatches?
What about error processing?  does the charset persist there as well?
How does someone remove the charset on the response's `Content-Type` after the fact now?
What happens if the user uses `HttpServletResponse.reset()` after the charset has been set?

**Comment by janbartel** [comment]
@joakime the change doesn't affect any of those things. All it does, is ensure that setContentType() without a char set doesn't change the existing value of the character encoding. As Request has setCharacterEncoding() method and a setContentType() method, the char set can be independent of the content-type. Previously we were treating the content-type and char-set as always inextricably linked, even if setContentType() didn't explicitly pass in a char-set (we used the defaults from MimeType).

File: jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java [hunk 1]
```diff
@@ -624,7 +635,7 @@ public void testResetContentTypeWithCharacterEncoding() throws Exception
 
         response.setContentType("wrong/answer;charset=utf-8");
         response.setContentType("foo/bar");
-        assertEquals("foo/bar", response.getContentType());
+        assertEquals("foo/bar;charset=utf-8", response.getContentType());
```
Comment (joakime) [inline jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:638]: Does this get sent over the wire on HTTP/1.1 as ...

```
Content-Type: foo/bar; charset=utf-8
```

?  

If so, then this PR is either doing too much, or we lack the extra tests to prove that we allow a user to turn off that `charset=` portion on the HTTP response.

**Review by joakime — COMMENTED** [review]
If this change makes the HTTP response include the `charset` line when we don't want it, then this is going to break many libraries (like REST / Jersey / Jax-RS).

File: jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java (hunk 1 above)
Comment (janbartel) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:638]: This change won't force a change to the charset: in fact, the change *prevents* a change to the charset that that spec does not support.  Unfortunately, the javadoc for ServletResponse.setContentType() does not make this behaviour clear in the case of the content-type without char-set, so all we have to go on is the tck test: https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/src/com/sun/ts/tests/servlet/api/common/response/ResponseTests.java#L680

Comment (joakime) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:638]: I created a new branch (from this one) called [`jetty-10.0.x-4719-keepcharencoding-extratests`](https://github.com/eclipse/jetty.project/tree/jetty-10.0.x-4719-keepcharencoding-extratests) with new testcases to show the concern (see commit e0b322f7d556837986accb3e881d5ec374285ede).

Two of the new testcases in ResponseHeadersTest are now failing ...

* testCharsetChangeToJsonMimeType()
* testCharsetChangeToJsonMimeTypeSetCharsetToNull()

> Interesting aside, I didn't know you could use `HttpServletResponse.reset()` after obtaining a `getWriter()` and still use that writer.

Comment (janbartel) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:638]: @joakime I've merged your branch into mine, and added a fix for content types like json that must have an assumed char set. Thanks for the concrete tests!!

**Review by joakime — APPROVED** [review]
LGTM

**Pull request merged**